### PR TITLE
inductor: support masked load for cpu path

### PR DESCRIPTION
**Pull request by XiaobingSuper**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107670

For max_pooling code:

```



#pragma GCC ivdep
                    for(long i2=static_cast<long>(0L); i2<static_cast<long>(56L); i2+=static_cast<long>(1L))
                    {
                        for(long i3=static_cast<long>(0L); i3<static_cast<long>(64L); i3+=static_cast<long>(16L))
                        {
                            auto tmp0 = at::vec::Vectorized<int>(static_cast<int>((-1L) + (2L*i1)));
                            auto tmp1 = at::vec::Vectorized<int>(static_cast<int>(0));
                            auto tmp2 = to_float_mask(tmp0 >= tmp1);
                            auto tmp3 = at::vec::Vectorized<int>(static_cast<int>(112));
                            auto tmp4 = to_float_mask(tmp0 < tmp3);
                            auto tmp5 = tmp2 & tmp4;
                            auto tmp6 = at::vec::Vectorized<int>(static_cast<int>((-1L) + (2L*i2)));
                            auto tmp7 = to_float_mask(tmp6 >= tmp1);
                            auto tmp8 = to_float_mask(tmp6 < tmp3);
                            auto tmp9 = tmp7 & tmp8;
                            auto tmp10 = tmp5 & tmp9;
                            auto tmp11 = [&]
                            {
                                auto tmp12 = at::vec::Vectorized<bfloat16>::loadu(in_ptr0 + static_cast<long>((-7232L) + i3 + (128L*i2) + (14336L*i1) + (802816L*i0)), 16);
                                                        load
                                auto tmp13 = cvt_lowp_fp_to_fp32<bfloat16>(tmp12);
                                                        
                                return tmp13;
                            }
                            ;
                            auto tmp14 = decltype(tmp11())::blendv(at::vec::Vectorized<float>(-std::numeric_limits<float>::infinity()), tmp11(), to_float_mask(tmp10));
```

the index of ```tmp12 ``` may be a correct index, such as ```i1=0, i2=0, i3=0```, the index is ```-7232L```, it is not a valid index. We may meet segmentation fault error when we call ```tmp11()```, the original behavior is that only the ```tmp10```(index check variable) is true, we can safely get the value, this PR will support masked_load to fixing this issue.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

File: test/inductor/test_cpu_repro.py [hunk 1]
```diff
@@ -1808,7 +1808,7 @@ def get_index():
     @patch("torch.cuda.is_available", lambda: False)
     def test_maxpool2d_cpu_only(self):
         for dtype in vec_dtypes:
-            input = torch.randn(10, 32, 20, 20, dtype=dtype).to(
+            input = torch.randn(26, 32, 112, 112, dtype=dtype).to(
```
Comment (jgong5) [inline test/inductor/test_cpu_repro.py:1811]: The increase of shape sizes is necessary to trigger the failures?

File: torch/_inductor/codegen/cpp.py [hunk 2]
```diff
@@ -1469,10 +1485,20 @@ def load(self, name: str, index: sympy.Expr):
...
         else:
-            line = f"at::vec::Vectorized<float>::loadu({loadbuf})"
+            line = (
+                f"masked_load({loadbuf}, {load_mask})"
+                if load_mask
+                else f"at::vec::Vectorized<float>::loadu({loadbuf})"
+            )
+
         if non_contiguous:
+            # TODO: support masked_load for non_contiguous path?
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:1501]: Are you going to handle this in this PR or a separate one? Maybe add an assertion here first to make sure we are not hitting the unexpected case?

File: torch/_inductor/codegen/cpp.py [hunk 3]
```diff
@@ -1469,10 +1485,20 @@ def load(self, name: str, index: sympy.Expr):
         elif is_mask:
             line = f"flag_to_float_vec({loadbuf})"
         elif dtype in DTYPE_LOWP_FP:
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:1487]: `uint8` with `loadu_one_fourth` also needs a masked load? https://github.com/pytorch/pytorch/pull/107670/files#diff-5ab7b0235e2076a5fc6629ba0b109208940f5b94f5c13babc3e0f87cf4fcec82R1483

File: torch/_inductor/codegen/cpp_prefix.h [hunk 4]
```diff
@@ -248,6 +248,38 @@ inline float flag_to_float_scalar(T src) {
...
+    return _mm256_mask_loadu_ps(zero_vec, mmask, src);
+# endif
+}
+
+template <typename T>
+inline at::vec::Vectorized<T> masked_load(const T* src, at::vec::Vectorized<float> mask) {
+# if defined(CPU_CAPABILITY_AVX512)
+  auto all_ones = _mm512_set1_epi32(0xFFFFFFFF);
+  auto mmask = _mm512_cmp_epi32_mask(_mm512_castps_si512(mask), all_ones, _MM_CMPINT_EQ);
+  auto zero = _mm256_set1_epi16(0);
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp_prefix.h:270]: This is an implementation specialized for 16-bit? Shall we declare the template function explicitly in that way?

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: test/inductor/test_cpu_repro.py (hunk 1 above)
Comment (XiaobingSuper) [reply test/inductor/test_cpu_repro.py:1811]: Yes, a smalled shape can't triger the eerror.

File: torch/_inductor/codegen/cpp.py (hunk 2 above)
Comment (XiaobingSuper) [reply torch/_inductor/codegen/cpp.py:1501]: we can handle it at next step.

File: torch/_inductor/codegen/cpp_prefix.h (hunk 4 above)
Comment (XiaobingSuper) [reply torch/_inductor/codegen/cpp_prefix.h:270]: updated.

File: torch/_inductor/codegen/cpp.py (hunk 3 above)
Comment (XiaobingSuper) [reply torch/_inductor/codegen/cpp.py:1487]: yes, added this dtype.

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: torch/_inductor/codegen/cpp_prefix.h [hunk 5]
```diff
@@ -248,6 +248,53 @@ inline float flag_to_float_scalar(T src) {
...
+  auto temp = _mm256_mask_loadu_epi16(zero, mmask, src);
+  return _mm512_inserti32x8(_mm512_castsi256_si512(temp), zero, 1);
+# else // AVX2
+  auto all_ones = _mm256_set1_epi32(0xFFFFFFFF);
+  auto mmask = _mm256_cmpeq_epi32(_mm256_castps_si256(mask), all_ones);
+  auto temp = _mm256_maskload_epi32(reinterpret_cast<const int32_t*>(src), mmask);
```
[the remark this answers is not in the document]
Comment (XiaobingSuper) [reply torch/_inductor/codegen/cpp_prefix.h:276]: code updated.

**Review by jgong5 — APPROVED** [review]

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**